### PR TITLE
Feature iteratordesignpattern

### PR DESCRIPTION
**Pull request by jeroenoverman**
Implemented two iterators for the ballgraph. 

-BallGraphBreadthFirstIterator:

This iterator iterates over all adjacent balls of the same color

-BallGraphAdjacentiterator:

Implements an Iterator that iterates over all the balls adjacent to a given ball

This iterator is currently only used by the other iterator

Makes the the ballgraph class less complex and improves readability. 

closes #95

### CODE REVIEW

File: core/src/com/group66/game/cannon/BallGraphAdjacentIterator.java [hunk 1]
```diff
@@ -57,5 +57,13 @@ public Ball next() {
...
         return ret;
     }
+    
+    /**
+     * remove is not supported by the iterator
+     */
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException();
+    }
```
Comment (11maxed11) [inline core/src/com/group66/game/cannon/BallGraphAdjacentIterator.java]: Just curious, what are those?

File: core/src/com/group66/game/cannon/BallGraphBreadthFirstConditionalIterator.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    /**
+     * returns the next object
+     * 
+     * @return The next object of the iterator
+     */
+    @Override
+    public Ball next() {
+        Ball ret = list.get(0);
```
Comment (ALobker) [inline core/src/com/group66/game/cannon/BallGraphBreadthFirstConditionalIterator.java]: Should there be a check if list still has elements?

File: core/src/com/group66/game/cannon/BallGraphAdjacentIterator.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * returns the next object
+     * 
+     * @return The next object of the iterator
+     */
+    @Override
+    public Ball next() {
+        Ball ret = list.get(0);
```
Comment (ALobker) [inline core/src/com/group66/game/cannon/BallGraphAdjacentIterator.java]: Should there be a check if list still has elements?

File: core/src/com/group66/game/cannon/BallGraphBreadthFirstConditionalIterator.java (hunk 2 above)
Comment (jeroenoverman) [inline core/src/com/group66/game/cannon/BallGraphBreadthFirstConditionalIterator.java]: The arraylist will throw an exception if it's out of bounds.

File: core/src/com/group66/game/cannon/BallGraphAdjacentIterator.java (hunk 1 above)
Comment (jeroenoverman) [inline core/src/com/group66/game/cannon/BallGraphAdjacentIterator.java]: In java 1.8 the function in the interfaced is changed in a default operation, so you don't have to override it. In java 1.7 it is not.

**Pull request merged**